packages/kit/src/runtime/client/parse.js

```diff
@@ -55,3 +56,22 @@ export function parse({ nodes, server_loads, dictionary, matchers }) {
 		return id === undefined ? id : [layouts_with_server_load.has(id), nodes[id]];
 	}
 }
+
+/**
+ * @param {import('types').CSRRouteServer} input
+ * @param {import('types').CSRPageNodeLoader[]} app_nodes Will be modified if a new node is loaded that's not already in the array
+ * @returns {import('types').CSRRoute}
+ */
+export function parse_server_route({ nodes, id, leaf, layouts, errors }, app_nodes) {
+	return {
+		id,
+		exec: () => {
+			throw new Error('shouldnt be called');
+		},
+		// By writing to app_nodes only when a loader at that index is not already defined,
+		// we ensure that loaders have referential equality when they load the same node.
+		errors: errors.map((n) => (n ? (app_nodes[n] ||= nodes[n]) : undefined)),
+		layouts: layouts.map((n) => (n ? [n[0], (app_nodes[n[1]] ||= nodes[n[1]])] : undefined)),
+		leaf: [leaf[0], (app_nodes[leaf[1]] ||= nodes[leaf[1]])]
```

Comment:
it feels like it shouldn't be necessary to continually update `app.nodes`, it's only used on startup AFAICT?

Reply:
No it's not, it's crucial to other logic where the client checks "is this a differen loader?" to see whether or not certain loaders are invalidated / need to be requested again. The comment above intended to hint at this but I guess it didn't make that sufficently clear?